doc/specs/#885 - Terminal Settings Model/Actions Addendum.md

```diff
@@ -0,0 +1,310 @@
+    - `ActionMap` must be responsible for changing `Command`s so that we can ensure `ActionMap` always has a correct internal state:
+        - It is important that `Command` only exposes getters (not setters) to ensure `ActionMap` is up to date.
+        - If a key chord is being changed, update the `_KeyMap` and the `Command` itself.
+        - If a key binding is being deleted, add an unbound action to the given key chord.
+    - This is similar to how color schemes are maintained today.
+    - In the event that name/key-chord is set to something that's already taken, we need to propagate those changes to
+       the rest of `ActionMap`. As we do with the JSON, we respect the last name/key-chord set by the user. See [Modifying Actions](#modifying-actions)
+       in potential issues.
+4. Serialization
+    - `Command::ToJson()` and `ActionMap::ToJson()` should perform most of the work for us. Simply iterate over the `_ActionMap` and call `Command::ToJson` on each action.
+    - See [Unbinding actions](#unbinding-actions) in potential issues.
+
+
+## UI/UX Design
+
+N/A
+
+## Capabilities
+
+N/A
+
+### Accessibility
+
+N/A
+
+### Security
+
+N/A
+
+### Reliability
+
+N/A
+
+### Compatibility
+
+N/A
+
+### Performance, Power, and Efficiency
+
+## Potential Issues
+
+### Layering Actions
+
+We need a way to determine where an action came from to minimize how many actions we serialize when we
+ write to disk. This is a two part approach that happens as we're loading the settings
+ 1. Load defaults.json
+     - For each of the actions in the JSON...
+         - Construct the `Command` (basically the `Command::LayerJson` we have today)
+         - Add it to the `ActionMap`
+           - this should update the internal state of `ActionMap` appropriately
+           - if the newly added key chord conflicts with a pre-existing one,
+              redirect `_KeyMap` to the newly added `Command` instead,
+              and update the conflicting one.
+2. Load settings.json
+     - Create a child for the `ActionMap`
+         - The purpose of a parent is to continue a search when the current `ActionMap` can't find a `Command` for a query. The parent is intended to be immutable.
+     - Load the actions array like normal into the child (see step 1)
+
+Introducing a parent mechanism to `ActionMap` allows it to understand where a `Command` came from. This allows us to minimize the number of actions we serialize when we write to disk, as opposed to serializing the entire list of actions.
+
+`ActionMap` queries will need to check their parent when they cannot find a matching `InternalActionID` in their `_ActionMap`.
+
+Since `NameMap` is generated upon request, we will need to pass a `std::set<InternalActionID>` as we generate the `NameMap` across each layer. This will ensure that each `Command` is only added to the `NameMap` once. We will start from the current layer and move up the inheritance tree to ensure that the current layer takes priority.

```

Comment:
Like, an internal API that's "get all commands" that returns a list with the child's commands first, then the parent's, then the parent's, ...  and then you iterate over that to build the map and exclusion  set

Reply:
We could, but if you use a list, we still need some way to keep track of which `InternalActionId`s were already added to the name map. So we would still need a `std::set` to ensure that's the case.

We could get the best of both worlds and store it to a `std::map<InternalActionID, Command>` then do what you said. It'd be something like an `accumulatedActionMap`, I guess? And that could be useful for debugging, if that's what you're thinking about.

Would that work?

Reply:
Offline discussion:
It would look something like this...
```cpp
std::vector<Command> _GetAllActions()
{
  std::vector<Command> actions;
  for (auto parent : _parents)
  {
    for (auto [_,cmd]: parent->_ActionMap)
    {
      actions.push_back(cmd);
    }
  }
  return actions;
}
// in generate name map
std::set<InternalActionID> visitedIDs{};
for(auto cmd : _GetAllActions())
{
  // check if cmd is in visitedIDs
  // if it is, skip it
  // otherwise, add it to the namemap
}
```

Could be useful for debugging and a bit clearer for anybody who comes along later and reads the code. Updating the spec.